fusion.ml

```diff
@@ -1119,30 +1150,41 @@ let rec type_subst i ty =
       |  _ -> failwith "Malformed term cannot be made into a construction"
 
   let rec constructionToTerm = function
-      | Comb(Comb(Const("QuoConst",_),strList),tyConv) -> Const(implode (readStringList strList),revTypeMatch tyConv)
+      | Comb(Comb(Const("QuoConst",_),strList),tyConv) ->
+          let con = (Const(implode(readStringList strList), revTypeMatch tyConv)) in
+          if properConst con then con  
+          else failwith "Not a proper construction" 
       | Comb(Comb(Const("QuoVar",_),strList),tyConv) -> Var(implode (readStringList strList),revTypeMatch tyConv)
       | Comb(Comb(Const("App",_),t1),t2) -> Comb(constructionToTerm t1,constructionToTerm t2)
       | Comb(Comb(Const("Abs",_),t1),t2) -> Abs(constructionToTerm t1,constructionToTerm t2)
       | Comb(Const("Quo",_),t) -> let ct = constructionToTerm t in Quote(ct,type_of ct)
       | other when type_of(other) = Tyapp("epsilon",[]) -> Hole(other,type_of other)
       | _ -> failwith "constructionToTerm"
 
-  let TERM_TO_CONSTRUCTION tm = match tm with
-      |  Quote(exp,t) when type_of exp = t -> Sequent([],safe_mk_eq tm (termToConstruction exp))
-      |  Quote(_,_) -> failwith "TERM_TO_CONSTRUCTION: BAD QUOTE"
-      | _ -> failwith "TERM_TO_CONSTRUCTION"
-  
-  let CONSTRUCTION_TO_TERM tm = try Sequent([],safe_mk_eq tm (mk_quote (constructionToTerm tm))) with Failure _ -> failwith "CONSTRUCTION_TO_TERM"
+  let is_proper_construc c = if can constructionToTerm c then true else false 
+
+  let proper_e_term tm = is_proper_construc tm || is_quote tm
+
+  let LAW_OF_QUO tm = match tm with
+    | Var(_,_) | Const(_,_) | Comb(_,_) | Abs(_,_) | Quote(_,_) | Hole(_,_) -> 
+      if is_eval_free tm then Sequent([],safe_mk_eq (mk_quote tm) (termToConstruction tm))
+      else failwith "Not eval free"
+    | Eval(_,_) -> failwith "Not eval free"
+    | _ -> failwith "Not a proper term"  
 
 
   (*These conversion functions can be used on their own but mainly will be used to construct tactics. They will search through a term for the first applicable instance and return the result of applying
   the relevant function to it*)
 
-  let rec TERM_TO_CONSTRUCTION_CONV tm = match tm with
-    | Const(a,b) -> failwith "TERM_TO_CONSTRUCTION_CONV"
-    | Quote(_,_) -> TERM_TO_CONSTRUCTION tm
-    | Comb(a,b) -> try TERM_TO_CONSTRUCTION_CONV a with Failure _ -> try TERM_TO_CONSTRUCTION_CONV b with Failure _ -> failwith "TERM_TO_CONSTRUCTION_CONV"
-    | _ -> failwith "TERM_TO_CONSTRUCTION_CONV"
+  let LAW_OF_QUO_CONV tm = LAW_OF_QUO tm
+
+  let rec QUOTE_TO_CONSTRUCTION_CONV tm = match tm with
```

Comment:
you can use `function` for this pattern, i.e.
```ocaml
let foo xx = match xx with
```
is just
```ocaml
let foo = function
```
(there are more places where this can be used)